### PR TITLE
Deprecate usage of IConfig as a singleton in Certbot

### PR DESCRIPTION
**Pull request by adferrand**
As discussed in #8835, this PR deprecates the usage of `config` as a singleton in Certbot. Concretely, it means that `zope.component.getUtility` is not used in the Certbot codebase anymore to retrieve the `config` object, which is passed as regular function argument instead.

The calls of `zope.component.provideUtility` for `config` are kept for compatibility, and appropriate deprecation warnings are emitted for third parties.

### CODE REVIEW

**Comment by adferrand** [comment]
Note that this PR changes the public API for `certbot.crypto_util.init_save_key` and `certbot.crypto_util.init_save_csr`). These changes are technically forward compatible because new parameters are added as named parameters with default values. These values are also the default one in the `config` object.

**Comment by bmw** [comment]
Thanks for working on this! I have two design changes I wanted to suggest:

1. Instead of monkey patching `zope.component`, what do you think about warning on accesses of `interfaces.IConfig` like I suggested at https://github.com/certbot/certbot/pull/8835#issuecomment-843396347? I think this probably shouldn't be added now and instead would be added as part of the `zope.interface` work. I personally like this approach better because I worry that monkey patching is somewhat fragile and think we should probably be warning about the interface going away anyway so it simplifies the code.
2. I think the API changes made to the two `crypto_util` functions is somewhat aggressive. I feel this way because the current configuration values are no longer respected and [strict_permissions actually defaults to `False`](https://github.com/certbot/certbot/blob/3640b8546e137916fd9e6c28b1ecb7b2faaffc1d/certbot/certbot/_internal/constants.py#L68). Instead, what do you think about either:

a) Making two new functions named something like `create_save_{csr,key}` that accepts parameters like you have here while `init_save_*` would raise deprecation warnings telling people to use the create functions, grab the values from the global config using `zope`, and then call the new functions.

b) Making the default values of the parameters to `init_save_*` `None` and if they are `None`, raise a warning saying these values must be set in a future release and grab the current value from the global config object.

What do you think?

**Comment by adferrand** [comment]
Indeed on the scope of `zope.interface`it makes more sense to emit warnings when the `certbot.interfaces` module is imported. My initial approach was to reuse this module for the ABC classes, making the deprecation  logic a little trickier to implement, but we can certainly circumvent that or use a new module name.

For the public API, I have a slight preference for solution b, because it avoids to add more content (functions) to the API, and the names `init_save_*` are quite appropriate and should be kept if possible.

**Comment by adferrand** [comment]
After some discussions, I prefer to go with option a for the public API, because the behavior induced by solution b on existing deprecated options is not obvious and harm usability.

File: certbot/certbot/crypto_util.py [hunk 1]
```diff
@@ -53,6 +54,8 @@ def init_save_key(
     :param str key_type: Key Type [rsa, ecdsa]
     :param str elliptic_curve: Name of the elliptic curve if key type is ecdsa.
     :param str keyname: Filename of key
+    :param boolean strict_permissions: If true, key will be saved with strict permissions
+        (POSIX mode 0600).
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:58]: I think a more accurate description here would be something like:

> If true and key_dir exists, an exception is raised if the directory doesn't have 0700 permissions or isn't owned by the current user.

File: certbot/certbot/crypto_util.py [hunk 2]
```diff
@@ -84,27 +86,61 @@ def init_save_key(
...
 
     :param privkey: Key to include in the CSR
     :type privkey: :class:`certbot.util.Key`
-
     :param set names: `str` names to include in the CSR
-
     :param str path: Certificate save directory.
+    :param boolean must_staple: If true, include the TLS Feature extension "OCSP Must Staple"
+    :param boolean strict_permissions: If true, the CSR file will be saved with strict
+        permissions (POSIX mode 0600).
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:133]: I think a more accurate description here would be something like:

> If true and path exists, an exception is raised if the directory doesn't have 0755 permissions or isn't owned by the current user.

File: certbot/certbot/_internal/client.py [hunk 3]
```diff
@@ -394,6 +396,7 @@ def _get_order_and_authorizations(self, csr_pem: str,
         """Request a new order and complete its authorizations.
 
         :param str csr_pem: A CSR in PEM format.
+        :param interfaces.IConfig config: current Certbot configuration
```
Comment (bmw) [inline certbot/certbot/_internal/client.py:399]:
```suggestion
```

File: certbot/certbot/crypto_util.py [hunk 4]
```diff
@@ -84,27 +86,61 @@ def init_save_key(
...
+                 must_staple: bool = False, strict_permissions: bool = True) -> util.CSR:
     """Initialize a CSR with the given private key.
 
     :param privkey: Key to include in the CSR
     :type privkey: :class:`certbot.util.Key`
-
     :param set names: `str` names to include in the CSR
-
     :param str path: Certificate save directory.
+    :param boolean must_staple: If true, include the TLS Feature extension "OCSP Must Staple"
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:131]:
```suggestion
    :param bool must_staple: If true, include the TLS Feature extension "OCSP Must Staple"
```

File: certbot/certbot/crypto_util.py [hunk 5]
```diff
@@ -84,27 +86,61 @@ def init_save_key(
...
+    :param boolean strict_permissions: If true, the CSR file will be saved with strict
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:132]:
```suggestion
    :param bool strict_permissions: If true, the CSR file will be saved with strict
```

File: certbot/certbot/crypto_util.py [hunk 6]
```diff
@@ -53,6 +54,8 @@ def init_save_key(
     :param str key_type: Key Type [rsa, ecdsa]
     :param str elliptic_curve: Name of the elliptic curve if key type is ecdsa.
     :param str keyname: Filename of key
+    :param boolean strict_permissions: If true, key will be saved with strict permissions
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:57]:
```suggestion
    :param bool strict_permissions: If true, key will be saved with strict permissions
```

File: certbot/certbot/crypto_util.py [hunk 7]
```diff
@@ -84,27 +86,61 @@ def init_save_key(
...
+def init_save_key(key_size, key_dir, key_type="rsa", elliptic_curve="secp256r1",
+                  keyname="key-certbot.pem"):
+    """Initializes and saves a privkey.
+
+    Inits key and saves it in PEM format on the filesystem.
+
+    .. note:: keyname is the attempted filename, it may be different if a file
+        already exists at the path.
+
+    .. deprecated:: 1.16.0
```
Comment (bmw) [inline certbot/certbot/crypto_util.py:99]: Nice! I never knew about this syntax. Thanks for adding this.

**Review by bmw — CHANGES_REQUESTED** [review]
My only comment other than my small documentation suggestions that I wrote inline is I think it'd be nice to provide simple unit tests of the `init_save*` functions. I was imagining something like asserting that a warning is raised and and a call is made to `generate_*` as we expect. You can probably talk me out of writing these tests if you really don't want to do it, but due to the dynamic nature of Python and the fact I think writing these tests should be pretty easy, I personally think it's worth it.

**Review by bmw — APPROVED** [review]
LGTM!

**Pull request merged**